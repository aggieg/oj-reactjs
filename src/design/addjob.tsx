import React from 'react';

type AddjobProps = {

    options: string[]; 

    hideForm: () => void;    

    launchDesignForm: (tag: string) =>  void;
}

type AddjobState = {

    selected: string | undefined;
}

export class AddJobForm extends React.Component<AddjobProps, AddjobState> {

    constructor(props: AddjobProps) {
        super(props);

        this.state = {

            selected: props.options[0]
        }

        this.handleChange = this.handleChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(e: React.FormEvent<HTMLFormElement>): void {
        e.preventDefault();
        if (this.state.selected) {
            this.props.launchDesignForm(this.state.selected);
        }
    }

    handleChange(event: React.FormEvent<HTMLSelectElement>): void {
        this.setState({
            selected: event.currentTarget.value
        });
    }

    render() {
        return <div className="oj-add-form-main">
            <form data-testid="add-form" onReset={this.props.hideForm} onSubmit={this.onSubmit}>
                <select value={this.state.selected} onChange={this.handleChange}>
                    {
                        this.props.options.map(e => <option key={e} value={e}>{e}</option>)
                    }
                </select>
                <button type="submit">OK</button>
                <button type="reset">Cancel</button>
            </form>
        </div>
    }

}
