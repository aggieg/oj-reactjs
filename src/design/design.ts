
export interface FormBuilder<T> {

    renderTextField(textField: TextField): T;

    renderSingleTypeSelection(singleTypeSelection: SingleTypeSelection): T;

    renderFieldGroup(fieldGroup: FieldGroup): T;
}

export interface FormItem {

    accept<T>(formBuilder: FormBuilder<T>): T;
}

interface DesignProperty extends FormItem {

}

class FormItemGroup {

    readonly items: Array<FormItem> = new Array();

    addItem(item: FormItem): FormItemGroup {
        this.items.push(item);
        return this;
    }

}

export class TextField implements DesignProperty {

    readonly property: string;

    readonly title: string;

    value?: string;

    constructor(property: string, title?: string) {
        this.property = property
        this.title = title ? title : property;
    }

    accept<T>(formBuilder: FormBuilder<T>): T {
        return formBuilder.renderTextField(this);
    }

}

export class DesignInstance extends FormItemGroup {

    readonly element: string;

    constructor(element: string) {
        super();
        this.element = element;
    }

}

export class SingleTypeSelection implements DesignProperty {

    readonly title: string

    readonly designFactory: TypeDesignFactory;

    private _instance?: DesignInstance;

    constructor(readonly property: string,
        readonly options: string[],
        designFactory: TypeDesignFactory,
        title?: string,
        instance?: DesignInstance) {

        this.designFactory = designFactory;
        this.title = title ? title : property;
        this._instance = instance;
    }

    get instance() {
        return this._instance;
    }

    change(element: string): void {
        if (element) {
            this._instance = this.designFactory.createDesign(new ElementOnly(element));
        }
        else {
            this._instance = undefined;
        }
    }

    accept<T>(formBuilder: FormBuilder<T>): T {

        return formBuilder.renderSingleTypeSelection(this);
    }

}

export class FieldGroup extends FormItemGroup implements FormItem {

    constructor(readonly title?: string) {
        super();
    }

    accept<T>(formBuilder: FormBuilder<T>): T {

        return formBuilder.renderFieldGroup(this);
    }
}

export enum ArooaType {
    Component,
    Value
}

interface Configuration {

    readonly element: string;

    getTextValue(property: string): string | undefined;

    getChild(property: string): Configuration | undefined;
}

class ElementOnly implements Configuration {

    readonly element: string;

    constructor(element: string) {
        this.element = element;
    }

    getTextValue(property: string): string | undefined {
        return undefined;
    }

    getChild(property: string): Configuration | undefined {
        return undefined;
    }

}

interface TypeDesignFactory {

    createDesign(configuration: Configuration): DesignInstance;
}

class TypeDesignFactoriesCache implements TypeDesignFactory {

    private readonly factoryByElement: Map<string, TypeDesignFactory> = new Map();

    createDesign(configuration: Configuration): DesignInstance {

        const { element } = configuration;

        const factory = this.factoryByElement.get(element)

        if (factory) {
            return factory.createDesign(configuration);
        }
        else {
            throw Error(`No factory for element ${element} `)
        }
    }

    addFactory(element: string, factory: TypeDesignFactory): void {

        this.factoryByElement.set(element, factory);
    }

}


export interface DesignFactory {

    createDesign(configuration: Configuration, arooaType: ArooaType): DesignInstance;
}

type FormItemFactory = (configuration: Configuration) => FormItem;


export class CachingDesignFactory implements DesignFactory {

    private readonly componentCache = new TypeDesignFactoriesCache();

    private readonly valueCache = new TypeDesignFactoriesCache();

    createDesign(configuration: Configuration, arooaType: ArooaType): DesignInstance {

        switch (arooaType) {
            case ArooaType.Component:
                return this.componentCache.createDesign(configuration)
            default:
                return this.valueCache.createDesign(configuration)
        }
    }

    formItemFactoryFrom(definition: any): FormItemFactory {

        let type: string = definition['@element'];
        if (!type) {
            throw new Error("No type.");
        }

        switch (type) {

            case 'design:field':
                {
                    const property: string = definition['property'];
                    const title: string = definition['title'];

                    return (configuration: Configuration) => {
                        const textField = new TextField(property, title ? title : property);
                        textField.value = configuration.getTextValue(property);
                        return textField;
                    }
                }
            case 'design:simple':
                {
                    const property: string = definition['property'];
                    const options: string[] = definition['options'];
                    const title: string = definition['title'];
                    const arooaType: ArooaType = definition['arooaType'];

                    const childFactory = arooaType === ArooaType.Component ?
                        this.componentCache : this.valueCache;

                    return (configuration: Configuration) => {

                        const childInstanceConfConf = configuration.getChild(property);
                        var childInstance;
                        if (childInstanceConfConf) {
                            childInstance = this.createDesign(childInstanceConfConf, arooaType);
                        }

                        const formItem = new SingleTypeSelection(property,
                            options,
                            childFactory,
                            title ? title : property,
                            childInstance
                        );

                        return formItem;
                    }
                }
            case 'design:group':
                {
                    const title = definition['title'];

                    let children: FormItemFactory[] = [];
                    if (definition['items']) {

                        const items: any[] = definition['items'];
                        children = items.map(this.formItemFactoryFrom);
                    }


                    return (configuration: Configuration) => {
                        const fieldGroup = new FieldGroup(title)
                        children.forEach(child => fieldGroup.addItem(child(configuration)));
                        return fieldGroup;
                    }
                }
            case 'design:tabs':
            case 'design:indexed':
            case 'design:mapped':        
            default: 'design:variable'
                throw new Error(`Unknown Form Item ${type}.`);
        }
    }

    addDesignDefinition(arooaType: ArooaType, definition: any): void {

        const element: string = definition['element'];

        const itemDefs = definition['form']['items'] as Array<any>;

        var formItemFactories: Array<FormItemFactory> = [];
        if (itemDefs) {
            formItemFactories = itemDefs.map(e => this.formItemFactoryFrom(e));
        }

        const designFactory = {
            createDesign(configuration: Configuration): DesignInstance {

                const instance = new DesignInstance(element);

                formItemFactories.forEach(factory => instance.addItem(factory(configuration)));

                return instance;
            }
        } as TypeDesignFactory;

        switch (arooaType) {
            case ArooaType.Component:
                this.componentCache.addFactory(element, designFactory);
                break;
            case ArooaType.Value:
                this.valueCache.addFactory(element, designFactory);
                break;
            default:
                throw new Error(`Unknown type ${arooaType}`);
        }
    }

    addDesignDefinitions(definitions: any) {

        const components: Array<any> = definitions['components'];
        const values: Array<any> = definitions['values'];

        if (components) {
            components.forEach(def => {
                this.addDesignDefinition(ArooaType.Component, def);
            });
        }

        if (values) {
            values.forEach(def => {
                this.addDesignDefinition(ArooaType.Value, def);
            });
        }
    }
}

export function configurationFromAny(obj: any): Configuration {

    const element: string = obj['@element'];
    if (!element) {
        throw new Error("No element in " + obj);
    }

    return {

        element: element,

        getTextValue: (property: string): string | undefined => {

            return obj[property];
        },

        getChild: (property: string): Configuration | undefined => {

            const child: any = obj[property];

            if (child) {
                return configurationFromAny(child);
            }
            else {
                return undefined;
            }
        }
    }
}

export function parse(instance: DesignInstance): any {

    const configuration: any = { '@element' : instance.element };

    class FooBuilder implements FormBuilder<void> {
    
        renderTextField(textField: TextField): void {
            if (textField.value) {
                configuration[textField.property] = textField.value;
            }
        }
    
        renderSingleTypeSelection(singleTypeSelection: SingleTypeSelection): void {
            if (singleTypeSelection.instance) {
                configuration[singleTypeSelection.property] = 
                    parse(singleTypeSelection.instance);
            }
        }
    
        renderFieldGroup(fieldGroup: FieldGroup): void {
            fieldGroup.items.forEach(e => e.accept(this));
        }

    }

    instance.items.forEach(item => item.accept(new FooBuilder()));

    return configuration;
}