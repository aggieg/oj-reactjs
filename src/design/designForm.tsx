import React from 'react';
import * as design from './design';
import { render } from '@testing-library/react';

function Field(props: { 
    formItem: design.FormItem;
    formBuilder: CompFormBuilder;
}) {

    const TheField = props.formItem.accept(props.formBuilder);

    return <TheField />
}

function FieldGroup(props: { 
    formItems: design.FormItem[];
    formBuilder: CompFormBuilder;
 }) {

    return (
        <ul className="oj-form-group">
            {
                props.formItems.map((item, index) =>
                    <Field key={index} formItem={item} formBuilder={props.formBuilder}/>)
            }
        </ul>
    )
}

class CompFormBuilder implements design.FormBuilder<React.ComponentClass<any, any>> {

    readonly level: number;

    constructor(level: number = 0) {
        this.level = level;
    }

    private fieldId(property: string): string {
        return `df_${property}_${this.level}`;
    }

    renderTextField(textField: design.TextField): React.ComponentClass<any, any> {

        const { title, property } = textField;
        const fieldId: string = this.fieldId(property);

        return class extends React.Component<any, any> {

            constructor(props: {}) {
                super(props);

                this.state = { value: textField.value }
                this.handleChange = this.handleChange.bind(this);
            }

            handleChange(event: React.FormEvent<HTMLInputElement>): void {
                const value = event.currentTarget.value;
                textField.value = value;
                this.setState({ value: textField.value });
            }

            render() {
                return (
                    <li>
                        <label htmlFor={fieldId}>{title}</label>
                        <input type="text" id={fieldId} value={this.state.value || ''}
                            onChange={this.handleChange} />
                    </li>
                );
            }
        };
    }

    renderSingleTypeSelection(singleTypeSelection: design.SingleTypeSelection): React.ComponentClass<any, any> {

        const fieldId = this.fieldId(singleTypeSelection.property);
        const title = singleTypeSelection.title;

        const level = this.level;

        return class extends React.Component<{}, any> {

            constructor(props: {}) {
                super(props);

                this.state = { selection: singleTypeSelection.instance?.element }
                this.handleChange = this.handleChange.bind(this);
            }

            handleChange(event: React.FormEvent<HTMLSelectElement>): void {
                const value = event.currentTarget.value;
                singleTypeSelection.change(value);
                this.setState({ selection: singleTypeSelection.instance?.element })
            }

            render() {

                let childForm;

                if (singleTypeSelection.instance) {
                    childForm = <InternalForm designInstance={singleTypeSelection.instance}
                        level={level + 1}/>
                }

                return (
                    <li>
                        <label htmlFor={fieldId}>{title}</label>
                        <select id={fieldId}
                            value={this.state.selection}
                            onChange={this.handleChange}>
                            {
                                singleTypeSelection.options.map(
                                    e => <option key={e} value={e}>{e}</option>)
                            }
                        </select>
                        { childForm }
                    </li>
                );
            }
        }
    }

    renderFieldGroup(fieldGroup: design.FieldGroup): React.ComponentClass<any, any> {

        const formBuilder = new CompFormBuilder(this.level + 1);

        return class extends React.Component<any, any> {

            render() {
                return (
                    <FieldGroup formItems={fieldGroup.items} formBuilder={formBuilder}/>
                );
            }
        };
    }
}

type DesignFormProps = {
    designInstance: design.DesignInstance;
    level?: number;
}

class InternalForm extends React.Component<DesignFormProps> {

    formBuilder: CompFormBuilder;

    constructor(props: DesignFormProps) {
        super(props);
        this.formBuilder = new CompFormBuilder(props.level ? props.level : 0);
    }

    render(): React.ReactNode {

        return (

            <div className="oj-design-form">
                <h4 className="oj-design-form-title">{this.props.designInstance.element}</h4>
                <FieldGroup formItems={this.props.designInstance.items} 
                        formBuilder={this.formBuilder}/>
            </div>
        )
    }

}

export class DesignForm extends React.Component<DesignFormProps> {

    constructor(props: DesignFormProps) {
        super(props);
    }

    render(): React.ReactNode {

        return (
            <div className="oj-design-form-main">
                <form>
                    <InternalForm designInstance={this.props.designInstance}/>
                </form>
            </div>
        )
    }

}

