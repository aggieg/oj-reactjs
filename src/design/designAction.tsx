
import React from 'react';
import ReactDOM from 'react-dom';
import { ActionFactory, ActionContext, Action, contextSearch } from '../menu/actions';
import { ConfigurationOwner } from '../remote/remote';
import { DesignModel, ParserDesignModel } from './design';
import { DesignForm } from './designForm';
import { definedOrError } from '../main/util';

export class DesignActionFactory implements ActionFactory {

    createAction(actionContext: ActionContext): Action | null {
 
        const configOwner: ConfigurationOwner | null = contextSearch(actionContext, ConfigurationOwner);

        if (configOwner == null) {
            return null;
        }

        return {

            name: "Design",

            isEnabled: true,

            perform: () => {

                configOwner.formFor(actionContext.proxy)
                    .then(formText => {
                        const designModel = new ParserDesignModel(
                            {
                                formConfiguration: JSON.parse(formText),
                                saveAction: (config: any) => { 
                                    console.log("Save TBD."); 
                                },
                                newForm: (element: string, isComponent: boolean, propertyClass: string) =>
                                    configOwner.blankForm(isComponent, element, propertyClass)
                            });

                        const formDiv = definedOrError(document.getElementById('ojForm'), "No form div.");
                        function hide() {
                            formDiv.style.display = 'block';
                            ReactDOM.unmountComponentAtNode(formDiv);
                        }

                        ReactDOM.render(
                            <DesignForm designModel={designModel}
                                hideForm={hide} />,
                            formDiv);

                        formDiv.style.display = 'block';
                    });
            }

        }
    }
}
