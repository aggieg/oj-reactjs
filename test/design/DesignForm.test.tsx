import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { DesignForm } from '../../src/design/designForm'
import { DesignInstance, TextField, SingleTypeSelection, CachingDesignFactory, configurationFromAny, ArooaType, parse } from '../../src/design/design';

const designDefinitions = require('./data/FruitDesigns.json');
const configuration = require('./data/FruitConfiguration.json');


test('Test Render Text Field', () => {
   
    const factories: CachingDesignFactory = new CachingDesignFactory();
    factories.addDesignDefinitions(designDefinitions);

    const designInstance = factories.createDesign(configurationFromAny(configuration),
        ArooaType.Component);
  
    const result = render(<DesignForm designInstance={ designInstance} />);

    expect(result).toMatchSnapshot();    

    const descriptionInput = result.getByLabelText("Description");

    fireEvent.change(descriptionInput, { target: { value: "An Afternoon Snack"} });

    const tasteInput = result.getByLabelText("Taste");

    fireEvent.change(tasteInput, { target: { value: "Tangy"} });

    expect(result).toMatchSnapshot();    

    const after: any = parse(designInstance);

    expect(after['description']).toBe('An Afternoon Snack');
    expect(after['fruit']['taste']).toBe('Tangy');
});

test('Test Render Design Instance', () => {

    const factories: CachingDesignFactory = new CachingDesignFactory();
    factories.addDesignDefinitions(designDefinitions);

    const designInstance = factories.createDesign(configurationFromAny(configuration),
        ArooaType.Component);
    

    const result = render(<DesignForm designInstance={ designInstance} />);

    const fruitSelection = result.getByLabelText("Fruit");

    fireEvent.change(fruitSelection, { target: { value: "orange"} });

    const seedlessInput = result.getByLabelText("Seedless");

    fireEvent.change(seedlessInput, { target: { value: "true"} });

    expect(result).toMatchSnapshot();    

    const after: any = parse(designInstance);

    expect(after['fruit']['@element']).toBe('orange');
    expect(after['fruit']['seedless']).toBe('true');
})