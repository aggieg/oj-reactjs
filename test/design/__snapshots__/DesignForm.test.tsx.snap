// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render Design Instance 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="oj-design-form-main"
      >
        <form>
          <div
            class="oj-design-form"
          >
            <h4
              class="oj-design-form-title"
            >
              snack
            </h4>
            <ul
              class="oj-form-group"
            >
              <li>
                <label
                  for="df_description_0"
                >
                  Description
                </label>
                <input
                  id="df_description_0"
                  type="text"
                  value=""
                />
              </li>
              <li>
                <label
                  for="df_fruit_0"
                >
                  Fruit
                </label>
                <select
                  id="df_fruit_0"
                >
                  <option
                    value="value"
                  >
                    value
                  </option>
                  <option
                    value="apple"
                  >
                    apple
                  </option>
                  <option
                    value="orange"
                  >
                    orange
                  </option>
                </select>
                <div
                  class="oj-design-form"
                >
                  <h4
                    class="oj-design-form-title"
                  >
                    orange
                  </h4>
                  <ul
                    class="oj-form-group"
                  >
                    <li>
                      <label
                        for="df_seedless_1"
                      >
                        Seedless
                      </label>
                      <input
                        id="df_seedless_1"
                        type="text"
                        value="true"
                      />
                    </li>
                  </ul>
                </div>
              </li>
            </ul>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="oj-design-form-main"
    >
      <form>
        <div
          class="oj-design-form"
        >
          <h4
            class="oj-design-form-title"
          >
            snack
          </h4>
          <ul
            class="oj-form-group"
          >
            <li>
              <label
                for="df_description_0"
              >
                Description
              </label>
              <input
                id="df_description_0"
                type="text"
                value=""
              />
            </li>
            <li>
              <label
                for="df_fruit_0"
              >
                Fruit
              </label>
              <select
                id="df_fruit_0"
              >
                <option
                  value="value"
                >
                  value
                </option>
                <option
                  value="apple"
                >
                  apple
                </option>
                <option
                  value="orange"
                >
                  orange
                </option>
              </select>
              <div
                class="oj-design-form"
              >
                <h4
                  class="oj-design-form-title"
                >
                  orange
                </h4>
                <ul
                  class="oj-form-group"
                >
                  <li>
                    <label
                      for="df_seedless_1"
                    >
                      Seedless
                    </label>
                    <input
                      id="df_seedless_1"
                      type="text"
                      value="true"
                    />
                  </li>
                </ul>
              </div>
            </li>
          </ul>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Text Field 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="oj-design-form-main"
      >
        <form>
          <div
            class="oj-design-form"
          >
            <h4
              class="oj-design-form-title"
            >
              snack
            </h4>
            <ul
              class="oj-form-group"
            >
              <li>
                <label
                  for="df_description_0"
                >
                  Description
                </label>
                <input
                  id="df_description_0"
                  type="text"
                  value=""
                />
              </li>
              <li>
                <label
                  for="df_fruit_0"
                >
                  Fruit
                </label>
                <select
                  id="df_fruit_0"
                >
                  <option
                    value="value"
                  >
                    value
                  </option>
                  <option
                    value="apple"
                  >
                    apple
                  </option>
                  <option
                    value="orange"
                  >
                    orange
                  </option>
                </select>
                <div
                  class="oj-design-form"
                >
                  <h4
                    class="oj-design-form-title"
                  >
                    apple
                  </h4>
                  <ul
                    class="oj-form-group"
                  >
                    <li>
                      <label
                        for="df_colour_1"
                      >
                        Colour
                      </label>
                      <input
                        id="df_colour_1"
                        type="text"
                        value="Green"
                      />
                    </li>
                    <ul
                      class="oj-form-group"
                    >
                      <li>
                        <label
                          for="df_taste_2"
                        >
                          Taste
                        </label>
                        <input
                          id="df_taste_2"
                          type="text"
                          value="Crispy"
                        />
                      </li>
                      <li>
                        <label
                          for="df_size_2"
                        >
                          Size
                        </label>
                        <input
                          id="df_size_2"
                          type="text"
                          value=""
                        />
                      </li>
                    </ul>
                  </ul>
                </div>
              </li>
            </ul>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="oj-design-form-main"
    >
      <form>
        <div
          class="oj-design-form"
        >
          <h4
            class="oj-design-form-title"
          >
            snack
          </h4>
          <ul
            class="oj-form-group"
          >
            <li>
              <label
                for="df_description_0"
              >
                Description
              </label>
              <input
                id="df_description_0"
                type="text"
                value=""
              />
            </li>
            <li>
              <label
                for="df_fruit_0"
              >
                Fruit
              </label>
              <select
                id="df_fruit_0"
              >
                <option
                  value="value"
                >
                  value
                </option>
                <option
                  value="apple"
                >
                  apple
                </option>
                <option
                  value="orange"
                >
                  orange
                </option>
              </select>
              <div
                class="oj-design-form"
              >
                <h4
                  class="oj-design-form-title"
                >
                  apple
                </h4>
                <ul
                  class="oj-form-group"
                >
                  <li>
                    <label
                      for="df_colour_1"
                    >
                      Colour
                    </label>
                    <input
                      id="df_colour_1"
                      type="text"
                      value="Green"
                    />
                  </li>
                  <ul
                    class="oj-form-group"
                  >
                    <li>
                      <label
                        for="df_taste_2"
                      >
                        Taste
                      </label>
                      <input
                        id="df_taste_2"
                        type="text"
                        value="Crispy"
                      />
                    </li>
                    <li>
                      <label
                        for="df_size_2"
                      >
                        Size
                      </label>
                      <input
                        id="df_size_2"
                        type="text"
                        value=""
                      />
                    </li>
                  </ul>
                </ul>
              </div>
            </li>
          </ul>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Text Field 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="oj-design-form-main"
      >
        <form>
          <div
            class="oj-design-form"
          >
            <h4
              class="oj-design-form-title"
            >
              snack
            </h4>
            <ul
              class="oj-form-group"
            >
              <li>
                <label
                  for="df_description_0"
                >
                  Description
                </label>
                <input
                  id="df_description_0"
                  type="text"
                  value="An Afternoon Snack"
                />
              </li>
              <li>
                <label
                  for="df_fruit_0"
                >
                  Fruit
                </label>
                <select
                  id="df_fruit_0"
                >
                  <option
                    value="value"
                  >
                    value
                  </option>
                  <option
                    value="apple"
                  >
                    apple
                  </option>
                  <option
                    value="orange"
                  >
                    orange
                  </option>
                </select>
                <div
                  class="oj-design-form"
                >
                  <h4
                    class="oj-design-form-title"
                  >
                    apple
                  </h4>
                  <ul
                    class="oj-form-group"
                  >
                    <li>
                      <label
                        for="df_colour_1"
                      >
                        Colour
                      </label>
                      <input
                        id="df_colour_1"
                        type="text"
                        value="Green"
                      />
                    </li>
                    <ul
                      class="oj-form-group"
                    >
                      <li>
                        <label
                          for="df_taste_2"
                        >
                          Taste
                        </label>
                        <input
                          id="df_taste_2"
                          type="text"
                          value="Tangy"
                        />
                      </li>
                      <li>
                        <label
                          for="df_size_2"
                        >
                          Size
                        </label>
                        <input
                          id="df_size_2"
                          type="text"
                          value=""
                        />
                      </li>
                    </ul>
                  </ul>
                </div>
              </li>
            </ul>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="oj-design-form-main"
    >
      <form>
        <div
          class="oj-design-form"
        >
          <h4
            class="oj-design-form-title"
          >
            snack
          </h4>
          <ul
            class="oj-form-group"
          >
            <li>
              <label
                for="df_description_0"
              >
                Description
              </label>
              <input
                id="df_description_0"
                type="text"
                value="An Afternoon Snack"
              />
            </li>
            <li>
              <label
                for="df_fruit_0"
              >
                Fruit
              </label>
              <select
                id="df_fruit_0"
              >
                <option
                  value="value"
                >
                  value
                </option>
                <option
                  value="apple"
                >
                  apple
                </option>
                <option
                  value="orange"
                >
                  orange
                </option>
              </select>
              <div
                class="oj-design-form"
              >
                <h4
                  class="oj-design-form-title"
                >
                  apple
                </h4>
                <ul
                  class="oj-form-group"
                >
                  <li>
                    <label
                      for="df_colour_1"
                    >
                      Colour
                    </label>
                    <input
                      id="df_colour_1"
                      type="text"
                      value="Green"
                    />
                  </li>
                  <ul
                    class="oj-form-group"
                  >
                    <li>
                      <label
                        for="df_taste_2"
                      >
                        Taste
                      </label>
                      <input
                        id="df_taste_2"
                        type="text"
                        value="Tangy"
                      />
                    </li>
                    <li>
                      <label
                        for="df_size_2"
                      >
                        Size
                      </label>
                      <input
                        id="df_size_2"
                        type="text"
                        value=""
                      />
                    </li>
                  </ul>
                </ul>
              </div>
            </li>
          </ul>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
